### PR TITLE
DOC: Add documentation related to X509_LOOKUPs

### PR DESCRIPTION
**Pull request by levitte**
Most of all, the base X509_LOOKUP functionality is now documented.
Furthermore, the names X509_LOOKUP_METHOD and X509_STORE are added for
reference.

### CODE REVIEW

File: doc/man3/X509_LOOKUP.pod [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+ X509_LOOKUP *X509_LOOKUP_new(X509_LOOKUP_METHOD *method);
+ int X509_LOOKUP_init(X509_LOOKUP *ctx);
+ int X509_LOOKUP_shutdown(X509_LOOKUP *ctx);
+ void X509_LOOKUP_free(X509_LOOKUP *ctx);
+
+ int X509_LOOKUP_set_method_data(X509_LOOKUP *ctx, void *data);
+ void *X509_LOOKUP_get_method_data(const X509_LOOKUP *ctx);
+
+ int X509_LOOKUP_ctrl(X509_LOOKUP *ctx, int cmd, const char *argc,
```
Comment (richsalz) [inline doc/man3/X509_LOOKUP.pod]: Do we use argc in many places elsewhere?  The dissonance with classic argc,argv is not good.

File: doc/man3/X509_LOOKUP.pod [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+X509_LOOKUP_set_method_data() and X509_LOOKUP_get_method_data()
+associates and retrieves a pointer to application data to and from the
+given B<X509_LOOKUP>.
+
+X509_LOOKUP_ctrl() is used to set or get additional data to or from a
+B<X509_LOOKUP> structure or its associated L<X509_LOOKUP_METHOD(3)>.
+The exact information to be set or received is determined by the
+I<cmd> number, and the various arguments I<argc>, I<argl> and I<*ret>
+hols the data being passed.  The I<cmd>s known to OpenSSL are
```
Comment (richsalz) [inline doc/man3/X509_LOOKUP.pod]: "hold" the data

File: doc/man3/X509_LOOKUP.pod [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+This can only be used with a lookup using the implementation
+L<X509_LOOKUP_file(3)>.
+
+X509_LOOKUP_add_dir() passes a directory specification from which
+certificates and CRLs are loaded on demand into the associated
+B<X509_STORE>.  I<type> indicates what type of object are expected.
+This can only be used with a lookup using the implementation
+L<X509_LOOKUP_hash_dir(3)>.
+
+X509_LOOKUP_add_store() passes a URI for a directory like structure
```
Comment (richsalz) [inline doc/man3/X509_LOOKUP.pod]: hyphenate directory-like

File: doc/man3/X509_STORE_add_cert.pod [hunk 4]
```diff
@@ -72,6 +78,11 @@ for the corresponding values used in certificate chain validation.  Their
 behavior is documented in the corresponding B<X509_VERIFY_PARAM> manual
 pages, e.g., L<X509_VERIFY_PARAM_set_depth(3)>.
 
+X509_STORE_add_lookup() finds or creates a L<X509_LOOKUP(3)> with the
+L<X509_LOOKUP_METHOD(3)> I<m> and adds it to the B<X509_STORE> I<v>.
```
Comment (richsalz) [inline doc/man3/X509_STORE_add_cert.pod]: m and v are bad names; at least change m to method.

**Comment by petrovr** [comment]
Richard Levitte wrote:
> Most of all, the base X509_LOOKUP functionality is now documented.
> Furthermore, the names X509_LOOKUP_METHOD
...
I cannot see anything rationale to keep low level methods to manage 
X.509 cache, i.e. X509 lookup.
With Store API there is no more need to implement X.509 lookup.
Store interface is quite generic and not limited to X.509 cache.
Lookup requires more complicated management.

So why to document lookup instead to deprecate?

Regards,
Roumen Petrov

File: doc/man3/X509_LOOKUP.pod [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+the lookup mechanism.
+
+X509_LOOKUP_init() initializes the given B<X509_LOOKUP>, after which
+it can be put to work.
+
+X509_LOOKUP_shutdown() tears down the given B<X509_LOOKUP>.
+
+X509_LOOKUP_free() destructs the given B<X509_LOOKUP>.
+
+X509_LOOKUP_set_method_data() and X509_LOOKUP_get_method_data()
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: Er, this documents existing stuff, which has been around for years. I have no interest in re-implementing any of this.

[the remark this answers is not in the document]
Comment (richsalz) [reply doc/man3/X509_LOOKUP.pod]: sure, that's reasonable.  maybe i'll do a PR to make that change if I have a spare hour.

**Comment by levitte** [comment]
> So why to document lookup instead to deprecate?

I was asked to document the add_store function, and didn't want to leave it half done. Of course, deprecation is an option, but we haven't gone there yet with these functions. There's already a lot going on...

**Comment by richsalz** [comment]
BTW, thanks for writing a bunch of documentation!

Perhaps ```X509_Lookup``` could have an intro paragraph that says what this does and when it's used?

**Comment by levitte** [comment]
> So why to document lookup instead to deprecate?

Getting back to this question, I am considering deprecating the file and hash directory X509_LOOKUP methods...  *But* I want to make sure I have full provided OSSL_STORE support in first.  Can you have the patience to wait and see, still?  It is an ongoing effort, but the basic KEYMGMT support has sucked up my time in the last month or so.

**Comment by levitte** [comment]
> Perhaps `X509_Lookup` could have an intro paragraph that says what this does and when it's used?

I would gladly see a suitable man7 page appear.  Not sure I'm the one to write it, though...  PR welcome!

File: doc/man3/X509_LOOKUP.pod [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+                                  X509_OBJECT *ret);
+ int X509_LOOKUP_by_fingerprint(X509_LOOKUP *ctx, X509_LOOKUP_TYPE type,
+                                const unsigned char *bytes, int len,
+                                X509_OBJECT *ret);
+ int X509_LOOKUP_by_alias(X509_LOOKUP *ctx, X509_LOOKUP_TYPE type,
+                          const char *str, i nt len, X509_OBJECT *ret);
+
+=head1 DESCRIPTION
+
+The B<X509_LOOKUP> structure is a collection of information needed to
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: Fixed

File: doc/man3/X509_LOOKUP.pod [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+look up certificates and CRLs according to an associated
+L<X509_LOOKUP_METHOD(3)>.  Multiple B<X509_LOOKUP> instances can be
+added to a L<X509_STORE(3)> to enable lookup in that store.
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: English confuses me on this...  I still haven't quite understood if "an" should be used with words that start with a written consonant or if it should be used with words that start with a vowel sound when spoken (such as X being pronounced "ex").  I often go by the latter rule, but sometimes get a pushback when doing so...

Anyway, confusion aside, I'm gladly following your advice on this.  Fixed

File: doc/man3/X509_LOOKUP.pod [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+X509_LOOKUP_new() creates a new B<X509_LOOKUP> with the implementation
+in I<method>.
+B<X509_LOOKUP>s can also be created by calling
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: Sure.  Done

File: doc/man3/X509_LOOKUP.pod (hunk 1 above)
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: Yeah I know, I had to look a few times before I could free myself of the argc,argv preconception...

However, in *this PR*, I'm focused on documenting what is.  If someone wants to change those parameter names, PR welcome :wink:

File: doc/man3/X509_LOOKUP.pod (hunk 2 above)
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: Fixed

File: doc/man3/X509_LOOKUP.pod (hunk 3 above)
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: Done

File: doc/man3/X509_STORE_add_cert.pod (hunk 4 above)
Comment (levitte) [reply doc/man3/X509_STORE_add_cert.pod]: This file is full of horrible names.  _ctx_ for a `X509_STORE`?  But that's what they're called in code...

But yeah, _m_ and _v_ are especially horrible.  Fixed

**Comment by levitte** [comment]
Ping!

**Comment by t8m** [comment]
@levitte missingcrypto.txt needs to be updated.

**Comment by levitte** [comment]
Thanks for the heads up.  I've fixed all nits I could find.

File: doc/man3/X509_LOOKUP.pod [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+X509_LOOKUP_set_method_data() and X509_LOOKUP_get_method_data()
+associates and retrieves a pointer to application data to and from the
+given B<X509_LOOKUP>.
+
+X509_LOOKUP_ctrl() is used to set or get additional data to or from a
+B<X509_LOOKUP> structure or its associated L<X509_LOOKUP_METHOD(3)>.
+The exact information to be set or received is determined by the
+I<cmd> number, and the various arguments I<argc>, I<argl> and I<*ret>
+holds the data being passed.  The I<cmd>s known to OpenSSL are
+discussed in more depth in L</Control Commands>.
```
Comment (mspncp) [inline doc/man3/X509_LOOKUP.pod]:
```suggestion
X509_LOOKUP_ctrl() is used to send a control command to the
B<X509_LOOKUP> structure or its associated L<X509_LOOKUP_METHOD(3)>.
The arguments of the control command are passed via I<argc> and I<argl>,
its return value via I<*ret>. The meaning of the arguments depends on the
I<cmd> number of the control command. In general, this function is
not called directly, but wrapped by a macro call, see below.
The control I<cmd>s known to OpenSSL are discussed in more depth
in L</Control Commands>.
```

File: doc/man3/X509_LOOKUP.pod [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+X509_LOOKUP_load_file() passes a filename to be loaded immediately
+into the associated B<X509_STORE>.  I<type> indicates what type of
+object is expected.
+This can only be used with a lookup using the implementation
+L<X509_LOOKUP_file(3)>.
+
+X509_LOOKUP_add_dir() passes a directory specification from which
+certificates and CRLs are loaded on demand into the associated
+B<X509_STORE>.  I<type> indicates what type of object are expected.
```
Comment (mspncp) [inline doc/man3/X509_LOOKUP.pod]:
```suggestion
B<X509_STORE>.  I<type> indicates what type of object is expected.
```

File: doc/man3/X509_LOOKUP.pod [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+B<X509_LOOKUP> using different criteria, where the looked up object is
+stored in I<ret>.  Some of the underlying B<X509_LOOKUP_METHOD>s will
+also cache objects matching the criteria in the associated
+B<X509_STORE>, which makes it possible to handle cases where the
+criteria have more than one hit.
+
+=head2 Control Commands
+
+The B<X509_LOOKUP_METHOD>s built into OpenSSL recognise the following
+X509_LOOKUP_ctrl() functions:
```
Comment (mspncp) [inline doc/man3/X509_LOOKUP.pod]:
```suggestion
The B<X509_LOOKUP_METHOD>s built into OpenSSL recognise the following
X509_LOOKUP_ctrl() I<cmd>s:
```

File: doc/man3/X509_LOOKUP.pod [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+=over 4
+
+=item B<X509_L_FILE_LOAD>
+
+This is the command that X509_LOOKUP_load_file() uses.  The filename
+is passed in B<argc>, and the type in <argl>.
```
Comment (mspncp) [inline doc/man3/X509_LOOKUP.pod]: Linebreak after uses.?

File: doc/man3/X509_LOOKUP.pod [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+
+=item B<X509_L_ADD_DIR>
+
+This is the command that X509_LOOKUP_add_dir() uses.  The directory
```
Comment (mspncp) [inline doc/man3/X509_LOOKUP.pod]: Same here and more locations.

File: doc/man3/X509_LOOKUP.pod [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+
+=head1 DESCRIPTION
+
+The B<X509_LOOKUP> structure holds the information needed to look up
+certificates and CRLs according to an associated L<X509_LOOKUP_METHOD(3)>.
+Multiple B<X509_LOOKUP> instances can be added to an L<X509_STORE(3)>
+to enable lookup in that store.
+
+X509_LOOKUP_new() creates a new B<X509_LOOKUP> with the implementation
+in I<method>.
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: That capital I looks misplaced...

File: doc/man3/X509_LOOKUP.pod [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+They can also be created by calling L<X509_STORE_add_lookup(3)>, which
+will associate a B<X509_STORE> with the lookup mechanism.
+
+X509_LOOKUP_init() initializes the given B<X509_LOOKUP>, after which
+it can be put to work.
+
+X509_LOOKUP_shutdown() tears down the given B<X509_LOOKUP>.
+
+X509_LOOKUP_free() destructs the given B<X509_LOOKUP>.
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: I wish I knew... If someone else has a suggestion, I'm all ears

[the remark this answers is not in the document]
Comment (richsalz) [reply doc/man3/X509_LOOKUP.pod]: I think you can only say that shutdown calls the implementation's shutdown function and close calls the close function and free's the allocated memory.

File: doc/man3/X509_LOOKUP.pod (hunk 14 above)
[the remark this answers is not in the document]
Comment (mspncp) [reply doc/man3/X509_LOOKUP.pod]: Should be an l

[the remark this answers is not in the document]
Comment (mspncp) [reply doc/man3/X509_LOOKUP.pod]: corrected

[the remark this answers is not in the document]
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: Applied

(note: it seems that github doesn't allow applying commit suggestions that are part of a multi-item review)

File: doc/man3/X509_LOOKUP.pod [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+X509_LOOKUP_set_method_data() and X509_LOOKUP_get_method_data()
+associates and retrieves a pointer to application data to and from the
+given B<X509_LOOKUP>.
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: Edited

File: doc/man3/X509_LOOKUP.pod (hunk 9 above)
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: I like.  Changed.

File: doc/man3/X509_LOOKUP.pod (hunk 11 above)
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: Yup

File: doc/man3/X509_LOOKUP.pod (hunk 12 above)
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: I believe we have a more lax attitude regarding that...  as long as we don't start bickering over one or two spaces again.

... but sure, why not

File: doc/man3/X509_LOOKUP.pod (hunk 15 above)
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: I think I've come up with something satisfactory.

File: doc/man3/X509_LOOKUP.pod (hunk 12 above)
Comment (mspncp) [reply doc/man3/X509_LOOKUP.pod]: Not because of the period-space-space stuff, but because the next sentence deserves to be on its own line.

Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: Oh, so you wanted an empty line in between? (pod translators process a paragraph at a time and do not respect line breaks unless in verbatim mode)

Comment (mspncp) [reply doc/man3/X509_LOOKUP.pod]: Good point. I'll take a look at the final manual page before answering your question...

Comment (mspncp) [reply doc/man3/X509_LOOKUP.pod]: Maybe better leave it as it is. The output looks very ragged with all the newlines.

But I noticed that in some locations you need to replace

     <argl>   ->  B<argl>

Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: You mean `I<argl>`

Comment (mspncp) [reply doc/man3/X509_LOOKUP.pod]: Why is it `B<argc>` and `I<argl>`?

Comment (levitte) [reply doc/man3/X509_LOOKUP.pod]: 'cause `B<argc>` is incorrect. I copied and pasted from somewhere and was sloppy, it seems.

**Comment by levitte** [comment]
Cleanup done (only of X509_LOOKUP.pod...  Similar cleanup of other files in another PR).  Please re-approve

**Comment by levitte** [comment]
Cherry-pick to 1.1.1 wasn't clean, so happens in a different PR: #11120

File: doc/man3/X509_LOOKUP.pod (hunk 7 above)
[the remark this answers is not in the document]
Comment (kaduk) [reply doc/man3/X509_LOOKUP.pod]: I know this change is long merged, but w.r.t. the general English usage question, I was always taught that "an" is used with words that start with a vowel sound when spoken.  This still is not an unambiguous rule, given (e.g.) our friends across the pond's propensity to voice initial aitches in some words, among other things.

**Pull request merged**